tools/mpremote: Improve soft restart mount reliability.

andrewleech opened the pull request:
I ran into an issue where on one of my esp32 boards with the new soft reset handling, the `if t - t_last_activity > QUIET_TIMEOUT:` check was being triggered early before `b">>> "` comes through so the remount doesn't happen.

On this board the `MPY: soft reboot` line comes through immediately (getting the routine past initial timeout) but then there's a slightly longer delay while the board restarts before it prints out the startup header and the repl prompt.

This PR adds some extra pattern monitoring during the timeout loop to track state if a soft restart is actually started.

An additional benefit of this is that before the change the full timeout was basically always being waited before mpremote continues. Now it generally exits earlier after just the (now shorter) QUIET_TIMEOUT period if soft reboot is not in progress, or QUIET_TIMEOUT long after `\n>` comes through.

pi-anl committed:
tools/mpremote: Improve soft restart mount reliability.

dpgeorge commented:
Unfortunately this doesn't work well when `sys.ps1` is not the default.  See #8449 which supports any prompt string, but note that PR does not fix the issue fixed by this PR.

andrewleech commented:
I think this one could be tweaked pretty easily to not depend as much upon the prompt. 
I'm already doing an earlier detection of `MPY: soft reboot` as a trigger to change the timeout behavior, that can probably be used as the primary key for whether the following remount code should be run.

After a soft reset ctrl-d the `MPY: soft reboot` line is printed pretty much immediately, then there can be a delay before the banner is printed. So that's the main point a longer timeout can be needed. 

So currently, I'm searching for `MPY: soft reboot` first, then a final line starting with `\n>`
Instead of looking for `\n>` it would probably be fine to look for the startup banner, then wait for inter-byte timeout to be exceeded to capture the rest.


The start of the banner can probably be detected with just `\nMicropython `, after which continue reading until there's no more bytes within a `QUIET_TIMEOUT = 0.1` period and continue from there.

dpgeorge commented:
I merged #8449.  When you get a chance, please update this on top of that commit, thanks!

andrewleech commented:
So.... this just got fixed, but a bit bigger :-)

I've rebased on current master and reworked it to monitor for the "soft reboot" line, then startup banner, then idle serial.read()

This now appears to work well in my testing for user modified prompts, remounting local every time.

Then I realised the other situation preventing remount is if the board has a startup main.py blocking interactive repl. When mpremote is first started on units like this it interrupts running code to inject the mount and present a repl.
I've attempted to do a similar thing here, detecting if soft-reset has started, but no banner is shown, interrupt the running code and wait for banner then serial.read() idle.

This does look a little ugly... but is functional? Would it be better to try to swallow/ignore the startup & interrupt text?
```
micropython/tools$ python3 ./mpremote/mpremote.py mount .
Local directory . is mounted at /remote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] to exit this shell
>
MicroPython v1.18-128-g2ea21abae on 2022-03-31; NUCLEO-WB55 with STM32WB55RGV6
Type "help()" for more information.
>>> 
>>> <HIT CRTL-D>
MPY: sync filesystems
MPY: soft reboot
2015-01-01 00:23:24.285 INFO root Bootloader status: NORMAL_STARTUP (0)
2015-01-01 00:23:24.506 INFO root Running main application v0.0.0-58-g6bc3336-dirty
2015-01-01 00:23:24.805 INFO blinky Starting Blinky task
2015-01-01 00:23:25.444 INFO BLE Advertising
Interrupt startup to remount local directory...
Traceback (most recent call last):
  File "main.py", line 30, in <module>
  File "main.py", line 26, in run
  File "device.py", line 61, in run
  File "uasyncio/core.py", line 1, in run_forever
  File "uasyncio/core.py", line 1, in run_until_complete
  File "uasyncio/core.py", line 1, in wait_io_event
  File "uasyncio/event.py", line 1, in ioctl
KeyboardInterrupt: 
MicroPython v1.18-128-g2ea21abae on 2022-03-31; NUCLEO-WB55 with STM32WB55RGV6
Type "help()" for more information.
>>> 
Remount local directory ./ at /remote
>>> 
```

pi-anl committed:
docs/mpremote: Add details about mount usage over soft-reboot.

dpgeorge commented:
Thanks for updating, merged in 79c05bd522addf14ddf188af712307ee76012d4a and 3b99b4d675f10a1679ee11e03410e2b50539bce2
